<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introducción" data-toc-modified-id="Introducción-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introducción</a></span></li><li><span><a href="#Setup" data-toc-modified-id="Setup-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Setup</a></span><ul class="toc-item"><li><span><a href="#Librerías" data-toc-modified-id="Librerías-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Librerías</a></span></li><li><span><a href="#Datos" data-toc-modified-id="Datos-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Datos</a></span></li></ul></li><li><span><a href="#Productos" data-toc-modified-id="Productos-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Productos</a></span><ul class="toc-item"><li><span><a href="#Limpieza-del-dataset" data-toc-modified-id="Limpieza-del-dataset-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Limpieza del dataset</a></span><ul class="toc-item"><li><span><a href="#Tratamientos-de-valores-nulos" data-toc-modified-id="Tratamientos-de-valores-nulos-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>Tratamientos de valores nulos</a></span></li></ul></li><li><span><a href="#Análisis" data-toc-modified-id="Análisis-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Análisis</a></span><ul class="toc-item"><li><span><a href="#Introducción" data-toc-modified-id="Introducción-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Introducción</a></span></li><li><span><a href="#Registro-de-clientes-por-fecha-de-ingesta" data-toc-modified-id="Registro-de-clientes-por-fecha-de-ingesta-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Registro de clientes por fecha de ingesta</a></span></li><li><span><a href="#Nivel-de-permanencia-de-los-clientes" data-toc-modified-id="Nivel-de-permanencia-de-los-clientes-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Nivel de permanencia de los clientes</a></span></li><li><span><a href="#Calcular-promedio-de-permanencia-por-número-de-meses" data-toc-modified-id="Calcular-promedio-de-permanencia-por-número-de-meses-3.2.4"><span class="toc-item-num">3.2.4&nbsp;&nbsp;</span>Calcular promedio de permanencia por número de meses</a></span></li><li><span><a href="#Número-de-clientes-(activos-y-no-activos)-por-fecha" data-toc-modified-id="Número-de-clientes-(activos-y-no-activos)-por-fecha-3.2.5"><span class="toc-item-num">3.2.5&nbsp;&nbsp;</span>Número de clientes (activos y no activos) por fecha</a></span></li><li><span><a href="#Número-de-clientes-activos,-no-activos-y-totales-por-producto-y-fecha" data-toc-modified-id="Número-de-clientes-activos,-no-activos-y-totales-por-producto-y-fecha-3.2.6"><span class="toc-item-num">3.2.6&nbsp;&nbsp;</span>Número de clientes activos, no activos y totales por producto y fecha</a></span></li><li><span><a href="#Valoración-de-los-clientes" data-toc-modified-id="Valoración-de-los-clientes-3.2.7"><span class="toc-item-num">3.2.7&nbsp;&nbsp;</span>Valoración de los clientes</a></span></li><li><span><a href="#Clientes-con-y-sin-productos-contratados" data-toc-modified-id="Clientes-con-y-sin-productos-contratados-3.2.8"><span class="toc-item-num">3.2.8&nbsp;&nbsp;</span>Clientes con y sin productos contratados</a></span><ul class="toc-item"><li><span><a href="#Productos-contratados" data-toc-modified-id="Productos-contratados-3.2.8.1"><span class="toc-item-num">3.2.8.1&nbsp;&nbsp;</span>Productos contratados</a></span></li><li><span><a href="#Productos-no-contratados" data-toc-modified-id="Productos-no-contratados-3.2.8.2"><span class="toc-item-num">3.2.8.2&nbsp;&nbsp;</span>Productos no contratados</a></span></li></ul></li><li><span><a href="#Productos-populares---Indice-de-contratación" data-toc-modified-id="Productos-populares---Indice-de-contratación-3.2.9"><span class="toc-item-num">3.2.9&nbsp;&nbsp;</span>Productos populares - Indice de contratación</a></span></li></ul></li><li><span><a href="#Visualización" data-toc-modified-id="Visualización-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Visualización</a></span><ul class="toc-item"><li><span><a href="#Nivel-de-permanencia---Bar" data-toc-modified-id="Nivel-de-permanencia---Bar-3.3.1"><span class="toc-item-num">3.3.1&nbsp;&nbsp;</span>Nivel de permanencia - Bar</a></span></li></ul></li></ul></li><li><span><a href="#Sociodemografía" data-toc-modified-id="Sociodemografía-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Sociodemografía</a></span><ul class="toc-item"><li><span><a href="#Limpieza-del-dataset" data-toc-modified-id="Limpieza-del-dataset-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Limpieza del dataset</a></span></li><li><span><a href="#Análisis" data-toc-modified-id="Análisis-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Análisis</a></span></li></ul></li><li><span><a href="#Actividad-comercial" data-toc-modified-id="Actividad-comercial-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Actividad comercial</a></span><ul class="toc-item"><li><span><a href="#Limpieza-del-dataset" data-toc-modified-id="Limpieza-del-dataset-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Limpieza del dataset</a></span></li><li><span><a href="#Análisis" data-toc-modified-id="Análisis-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Análisis</a></span></li></ul></li><li><span><a href="#Anexo" data-toc-modified-id="Anexo-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Anexo</a></span><ul class="toc-item"><li><span><a href="#Cuantiles" data-toc-modified-id="Cuantiles-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Cuantiles</a></span></li></ul></li><li><span><a href="#Referencias" data-toc-modified-id="Referencias-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Referencias</a></span></li></ul></div>

<center><span style="background-color: rgb(251, 160, 38); font-size: 32px;">  <b>NUCLIO DIGITAL SCHOOL -</b> MASTER EN DATA SCIENCE  </span></center>
    
<br>
    
<center><a href = https://nuclio.school/wp-content/uploads/2019/10/nucleoDS-newBlack.png > <img src="https://nuclio.school/wp-content/uploads/2019/10/nucleoDS-newBlack.png" width=400 height=100><a/></center>
    
<hr style="height:5px;border-width:0;color:orange;background-color:orange">
<center> <span style="font-size: 26px;">  <b>Proyecto Final de Máster - EasyMoney </span> </center>

+ Sesion: **Análisis de datos**
+ Bloque: **Trabajo Fin de Máster**
+ Curso: **Data Science Madrid 0521**
+ Alumno 1: **Sergio Masa Avís**
+ Alumno 2: **María Andreu**

<br>

<hr style="height:5px;border-width:0;color:orange;background-color:orange">


## Introducción

<img src="img/tarea_1_analisis.png" width="600" height="300">


<hr style="height:5px;border-width:0;color:orange;background-color:orange">

## Setup

### Librerías

In [2]:
# Librerías
import pandas as pd
import numpy as np
import plotly 
import plotly.graph_objects as go
from tqdm import tqdm

# Paquetes
from modules import analisis as anl
from modules import visualization as vs

# Configuración de librerías y paquetes
tqdm.pandas()
%matplotlib inline
%load_ext autoreload
%autoreload 2
plotly.offline.init_notebook_mode(connected = True)
pd.options.display.float_format = "{:.3f}".format


The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


### Datos

En el directorio data/ se encuentran tres fuentes de información almacenadas en formato csv:

- products_df
- sociodemographic_df
- commercial_activity

A continuación, se muestra la descripción proporcionada de los datos proporcionada por el Director de IT de EasyMoney

<img src="img/data_description.png" width="800" height="400">

<hr style="height:5px;border-width:0;color:orange;background-color:orange">

## Productos

__Descripción de variables del dataset.__

A continuación, se analizará el fichero "products_df.csv"

| Variable | Descripción |
| --- | --- |
| pk_cid | Identificador de cliente |
| pk_partition | Fecha de ingesta de los datos |
| credit_card | Tarjetas de crédito |
| debit_card | Tarjetas de débito |
| em_acount | Cuenta easyMoney |
| em_account_p | Cuenta easyMoney+ |
| em_account_pp | Cuenta easyMoney++ |
| emc_account | Cuenta easyMoney Crypto |
| funds | Fondos de inversión |
| loans | Préstamos |
| long_term_deposit | Depósitos a largo plazo |
| securities | Valores |
| short_term_deposit | Depósitos a corto plazo |
| mortgage | Mortgage |
| payroll | Domiciliaciones |
| payroll_account | Cuenta bonificada por domiciliaciones |
| pension_plan | Plan de pensiones |

In [ ]:
# Cargar datos
prod = anl.get_products_df(filename_dir='data/products_df.csv')

In [ ]:
# Mostrar un conjunto aleatorio de 10 muestras del dataset
prod.sample(10)

In [ ]:
lista_productos = anl.listar_productos(df_prod = prod)
pd.Series(lista_productos).to_csv('results/lista_productos.csv', sep=';', index=False, encoding='utf-8-sig')

In [ ]:
anl.resume_products_df(df_prod = prod)

<hr style="height:5px;border-width:0;color:orange;background-color:orange">

### Limpieza del dataset

#### Tratamientos de valores nulos

Como se puede observar en el resumen anterior, el dataset proporcionado por EasyMoney
de products tiene 61 valores nulos tanto para el producto "payroll" como para "pension_plan".

Dado que se trata de un dataset de clientes, se entiende que los valores nulos indican que dicho cliente no tiene activado ("dado de alta") dicho producto en el momento temporal correspondiente. Dicho de otra forma:

- Continúa siendo cliente dado que aparece en el registro.

- No tiene activado el producto y se establece el valor nulo a 0.

In [ ]:
# Visualización valores nulos
prod.loc[(prod['payroll'].isna()) | (prod['pension_plan'].isna())]

In [ ]:
columnas_con_nulos = ['payroll', 'pension_plan']
df_nulos = prod[columnas_con_nulos].copy()

prod[columnas_con_nulos] = anl.nulos_productos(df_prod = df_nulos.copy())

del df_nulos

In [ ]:
anl.resume_products_df(df_prod = prod)

<hr style="height:5px;border-width:0;color:orange;background-color:orange">

### Análisis

#### Introducción

La primera pregunta a la cuál es necesario definir una respuesta __¿que es un cliente y que tipos de clientes existen?__


Los clientes tiene un identificador ("pk_cid") que permite distinguir las características de cada uno. El identificador, la fecha de ingesta y el valor del cliente en un mes determinado ("id_type") para un producto proporciona información que puede ser de utilidad. Por lo tanto, se concluye que:

- Los identificadores ("pk_cid_type") pueden ser:

    - __conocido__. Identificadores que aparecen en algún mes anterior.
    
    - __nuevo__. No han aparecido anteriormente en el registro.


- El valor del cliente en un mes determinado para un producto puede tomar el valor 0 o 1, es decir, tiene dos estados. Dado que se trata de un dataset de clientes se clasifican como "tipo_identificador" en:

    - __nuevo_activo__. El cliente es nuevo y el valor del producto es igual a 1.
    
    - __nuevo_inactivo__. El cliente es nuevo y el valor del producto es igual a 0.
    
    - __conocido_inactivo__. El cliente es conocido y el valor del producto es 0.
    
    - __conocido_activo__. El cliente es conocido y el valor del producto es 1.
    

#### Registro de clientes por fecha de ingesta

En primer lugar, se creará un nuevo dataset que registre para todos los clientes mes a mes si:

- Aparece en el registro (valor igual a 1).
- No aparece en el registro (valor igual a 0).


El dataframe que se crea a continuación, tiene:

- Como index los identificadores únicos de los clientes.
- Como columnas cada una de las particiones en las cuáles se realizaron la ingesta de los datos.

In [ ]:
registro = anl.registro_clientes(df_prod=prod)
registro.to_csv('results/registro.csv', sep=';', index=True, encoding='utf-8-sig')

In [ ]:
registro.sample(10).T

#### Nivel de permanencia de los clientes


El dataset está compuesto por 17 fechas de ingesta. Por lo tanto, el número máximo de meses que un usuario puede estar como cliente en la base de datos es de 17.

El nivel de permanencia de un cliente se define como:

\begin{align}
n_{c} = \frac{mc_{c}}{t}
\end{align}

siendo:

- $n_{c}$. Nivel de permanencia del cliente.

- $mc_{c}$. Puntuación del cliente. Número de meses que ha pasado como cliente.

- $t$. Período de tiempo de evaluación en meses.

In [ ]:
periodo = registro.shape[1] # 17 meses

In [ ]:
nivel_permanencia = anl.nivel_permanencia(registro = registro, periodo = periodo)

In [ ]:
nivel_permanencia.sample(10)

In [ ]:
vs.violin_plot(y=nivel_permanencia['numero_meses'], title='Probabilidad de permanencia', name='numero_meses')

__En el violinplot se aprecian las siguientes cuatro regiones:__

- __(1) Fidelización baja.__ Aquellos clientes que permanencen en el registro entre 0 y 6 meses.
- __(2) Fidelización media.__ Aquellos clientes que permanecen en el registro entre 7 y 10 meses.
- __(3) Fidelización alta.__ Aquellos clientes que permanecen en el registro entre 11 y 15 meses.
- __(4) Fidelización completa.__ Aquellos clientes que permanencen en el registro entre 16 y 17 meses.


Por lo tanto, a continuación se identificarán en el dataset de nivel de permanencia estas cuatro regiones

In [ ]:
nivel_permanencia['region_permanencia'] = anl.region_permanencia(nivel_permanencia)
nivel_permanencia.to_csv('results/nivel_permanencia.csv', sep=';', index=False, encoding='utf-8-sig')

In [ ]:
nivel_permanencia

#### Calcular promedio de permanencia por número de meses

In [ ]:
promedio_permanencia = anl.promedio_permanencia(nivel_permanencia['numero_meses'])
promedio_permanencia.to_csv('results/promedio_permanencia.csv', sep=';', index=False, encoding='utf-8-sig')

In [ ]:
promedio_permanencia

#### Número de clientes (activos y no activos) por fecha


In [ ]:
# Clientes activos y no activos
anl.clientes_fecha(registro,show_msg=True)

#### Número de clientes activos, no activos y totales por producto y fecha

In [ ]:
# Tomar información de clientes con producto y 
clientes_activos, clientes_no_activos = anl.clientes_producto(df_prod=prod)

# Almacenar en csv
clientes_activos.to_csv('results/clientes_activos.csv', sep=';', index=True, encoding='utf-8-sig')
clientes_no_activos.to_csv('results/clientes_no_activos.csv', sep=';', index=True, encoding='utf-8-sig')

In [ ]:
clientes_activos

In [ ]:
clientes_no_activos

__Analizamos el dataframe__

In [ ]:
cp_informacion = anl.analizar_productos_clientes(df_prod=prod.copy(), ca=clientes_activos.copy())
cp_informacion.to_csv('results/informacion_clientes_con_producto.csv', sep=';', index=True, encoding='utf-8-sig')
cp_informacion

In [ ]:
cp_informacion.describe().T

#### Valoración de los clientes

In [ ]:
valoracion_clientes = anl.valorar_productos_clientes(df=prod, lista_productos = lista_productos)

# Almacenar dataset en csv (evitar tiempos de espera)
print('Almacenando valoración...')
valoracion_clientes.to_csv('results/valoracion_clientes.csv', sep=';', index=False, encoding='utf-8-sig')
valoracion_clientes

In [ ]:
# Lectura
valoracion_clientes = anl.get_valoracion_clientes(filename='results/valoracion_clientes.csv')
valoracion_clientes

#### Clientes con y sin productos contratados

Se distinguen dos tipos de clientes:
- Clientes con producto contratado en algún mes que producen una facturación en EasyMoney.
- Clientes sin producto contratado ningún mes que no producen una facturación (captados pero no facturan).

In [ ]:
productos_contratados, productos_no_contratados = anl.tipos_clientes(valoracion_clientes)

##### Productos contratados

In [ ]:
productos_contratados.sample(10)

In [ ]:
productos_contratados.describe().T

##### Productos no contratados

In [ ]:
productos_no_contratados.sample(10)

In [ ]:
productos_no_contratados.describe().T

#### Productos populares - Indice de contratación


En primer lugar, se desea conocer cuales son los productos con mayor indice de contratacion. 

Dado que un cliente puede tener como valor en un producto determinado 0 (servicio no activo) o 1 (servicio activo), por ejemplo, la media de contratación de un cliente durante los meses disponibles indica la proporción de servicios contratados en un determinado período. Por lo tanto, se calculará:

- La media general de todas las particiones de contratación (más cercano a 1 indica un mayor índice de popularidad general).


- La media por mes de ingesta de contratación (para tratar de identificar el crecimiento/decrecimiento de un producto determinado.


- La media por año de ingesta de contratación (para tratar de identificar el crecimiento/decrecimiento de un producto determinado.

In [ ]:
ind_contratacion = anl.indice_contratacion(df_prod=prod)

In [ ]:
ind_contratacion

### Visualización

#### Nivel de permanencia - Bar

In [ ]:
df_bar = pd.DataFrame(nivel_permanencia['region_permanencia'].value_counts(normalize=True))
df_bar = df_bar.reset_index().rename({'region_permanencia':'total', 'index':'region_permanencia'},axis=1)
df_bar.sort_values('total', ascending=False, inplace=True)


# Añadir columna de región
region = {
    'baja': '[1-6]',
    'media':'[7-10]',
    'alta': '[11-15]',
    'completa': '[16-17]'
}
df_bar['region'] = df_bar['region_permanencia'].map(region)
df_bar

<hr style="height:5px;border-width:0;color:orange;background-color:orange">

## Sociodemografía


__Descripción de variables del dataset.__

A continuación, se analizará el fichero "sociodemographic_df.csv"

| Variable | Descripción |
| --- | --- |
| pk_cid | Identificador de cliente |
| pk_partition | Fecha de ingesta de los datos |
| country_id | País de residencia del cliente |
| region_code | Provincia de residencia del cliente |
| gender | Sexo del cliente |
| age | Edad del cliente |
| deceased | ¿Cliente fallecido (N/S)? |
| salary | Ingresos brutos de la unidad familiar |

In [3]:
# Cargar datos
sociodemographic = anl.get_sociodemographic_df('data/sociodemographic_df.csv')
sociodemographic

pk_cid pk_partition country_id region_code gender  age deceased  \
0        1375586   2018-01-28         ES        29.0      H   35        N   
1        1050611   2018-01-28         ES        13.0      V   23        N   
2        1050612   2018-01-28         ES        13.0      V   23        N   
3        1050613   2018-01-28         ES        50.0      H   22        N   
4        1050614   2018-01-28         ES        50.0      V   23        N   
...          ...          ...        ...         ...    ...  ...      ...   
5962919  1166765   2019-05-28         ES        50.0      V   22        N   
5962920  1166764   2019-05-28         ES        26.0      V   23        N   
5962921  1166763   2019-05-28         ES        50.0      H   47        N   
5962922  1166789   2019-05-28         ES        50.0      H   22        N   
5962923  1550586   2019-05-28         ES        28.0      H   37        N   

            salary  
0        87218.100  
1        35548.740  
2       122179.110  
3       119775.540  
4              nan  
...            ...  
5962919  43912.170  
5962920  23334.990  
5962921        nan  
5962922 199592.820  
5962923        nan  

[5962924 rows x 8 columns]

In [10]:
pd.options.display.float_format = "{:.5f}".format

anl.resume_sociodemographic_df(df_sd=sociodemographic)

Tamaño del dataset
	Samples: 5962924
	Features: 8



Número de particiones del histórico: 17
Número de clientes únicos: 456373



Información del DataFrame
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5962924 entries, 0 to 5962923
Data columns (total 8 columns):
 #   Column        Dtype         
---  ------        -----         
 0   pk_cid        object        
 1   pk_partition  datetime64[ns]
 2   country_id    object        
 3   region_code   object        
 4   gender        object        
 5   age           int64         
 6   deceased      object        
 7   salary        float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(5)
memory usage: 363.9+ MB
None



Descripción del dataset
                age         salary
count 5962924.00000  4450821.00000
mean       29.76042   115816.72370
std        11.98038   199551.88905
min         2.00000     1202.73000
25%        22.00000    61500.63000
50%        25.00000    88654.65000
75%        34.00000   131669.910

### Limpieza del dataset

### Análisis

<hr style="height:5px;border-width:0;color:orange;background-color:orange">

## Actividad comercial

In [32]:
df_comm = anl.get_commercial_activity_df(filename_dir='data/commercial_activity_df.csv')
df_comm

pk_cid pk_partition entry_date entry_channel  active_customer  \
0        1375586   2018-01-28 2018-01-12           KHL          1.00000   
1        1050611   2018-01-28 2015-08-10           KHE          0.00000   
2        1050612   2018-01-28 2015-08-10           KHE          0.00000   
3        1050613   2018-01-28 2015-08-10           KHD          0.00000   
4        1050614   2018-01-28 2015-08-10           KHE          1.00000   
...          ...          ...        ...           ...              ...   
5962919  1166765   2019-05-28 2016-08-14           KHE          0.00000   
5962920  1166764   2019-05-28 2016-08-14           KHE          0.00000   
5962921  1166763   2019-05-28 2016-08-14           KHE          1.00000   
5962922  1166789   2019-05-28 2016-08-14           KHE          0.00000   
5962923  1550586   2019-05-28 2019-05-13           NaN          1.00000   

                    segment  
0         02 - PARTICULARES  
1        03 - UNIVERSITARIO  
2        03 - UNIVERSITARIO  
3        03 - UNIVERSITARIO  
4        03 - UNIVERSITARIO  
...                     ...  
5962919  03 - UNIVERSITARIO  
5962920  03 - UNIVERSITARIO  
5962921   02 - PARTICULARES  
5962922  03 - UNIVERSITARIO  
5962923                 NaN  

[5962924 rows x 6 columns]

In [35]:
anl.resume_commercial_activity_df(df_comm=df_comm)

Tamaño del dataset
	Samples: 5962924
	Features: 6



Número de particiones del histórico: 17
Número de fechas de entrada de los clientes: 1497
Número de clientes únicos: 456373



Información del DataFrame
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5962924 entries, 0 to 5962923
Data columns (total 6 columns):
 #   Column           Dtype         
---  ------           -----         
 0   pk_cid           object        
 1   pk_partition     datetime64[ns]
 2   entry_date       datetime64[ns]
 3   entry_channel    object        
 4   active_customer  float64       
 5   segment          object        
dtypes: datetime64[ns](2), float64(1), object(3)
memory usage: 273.0+ MB
None



Descripción del dataset
       active_customer
count    5962924.00000
mean           0.40338
std            0.49058
min            0.00000
25%            0.00000
50%            0.00000
75%            1.00000
max            1.00000



Valores nulos

pk_cid                  0
pk_partition            0
entr

### Limpieza del dataset

### Análisis

<hr style="height:5px;border-width:0;color:orange;background-color:orange">

## Anexo

### Cuantiles

El término cuantil fue usado por primera vez por Kendall en 1940 son puntos tomados a intervalos regulares de la función de distribución de una variable aleatoria y suelen usarse por grupos que dividen la distribución en partes iguales. Estas partes se entienden como intervalos que comprenden la misma proporción de valores


De esta forma, el cuantil de orden $p$ de una distribución (con $0 < p < 1$) es el valor de la variable $x_{p}$ que marca un corte de modo que una proporción p de valores de la población es menor o igual que $x_{p}$. 

Por ejemplo:

- El cuantil de orden 0.36 dejaría un 36% de valores por debajo.

- El cuantil de orden 0.50 se corresponde con la mediana de la distribución.

Los más usados utilizados son:

- Cuartiles. Dividen a la distribución en cuatro partes (corresponden a los cuantiles 0.25, 0.50 y 0.75).


- Quintiles. Dividen a la distribución en cinco partes (corresponden a los cuantiles 0.20, 0.40; 0.60 y 0,80).


- Deciles. Dividen a la distribución en diez partes.


- Percentiles. Dividen a la distribución en cien partes.

En el cálculo de cuantiles con distribuciones de variable continua (por ejemplo, con datos agrupados) puede conseguirse fácilmente que las partes en que se divide la distribución sean exactamente iguales. Sin embargo, en las distribuciones de variable discreta (como el caso de datos aislados) estas serán aproximadamente iguales. 

La función cuantil asigna a cada p el punto de corte $x_{p}$, es decir, el valor del cuantil de orden p.

<hr style="height:5px;border-width:0;color:orange;background-color:orange">

## Referencias

[1] Pandas Groupby: Summarising, Aggregating, Grouping in Python
[url: shanelynn.ie/summarising-aggregation-and-grouping-data-in-python-pandas/]

[2] Quantile and Decile rank of a column in Pandas-Python [url: https://www.geeksforgeeks.org/quantile-and-decile-rank-of-a-column-in-pandas-python/]